src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py

```diff
@@ -0,0 +1,451 @@
+    DotProduct2CodeTrans, _get_array_bound
+from psyclone.tests.utilities import Compile
+
+
+# Utilities
+
+def check_validate(code, expected, fortran_reader, index=0):
+    '''Utility function that takes Fortran code that is unsupported by the
+    DotProduct2CodeTrans transformation and checks that the expected
+    exception is raised when the validate function is applied.
+
+    :param str code: the tangent linear code stored in a string.
+    :param str expected: the expected exception text.
+    :param fortran_reader: converts Fortran into PSyIR.
+    :type fortran_reader: :py:class:`psyclone.psyir.frontend.fortran`
+    :param integer index: specifies which binaryoperation to check in \
+        code order. Defaults to 0 (which is the first).
+
+    '''
+    psyir = fortran_reader.psyir_from_source(code)
+    dot_product = psyir.walk(BinaryOperation)[index]
+    assert dot_product.operator == BinaryOperation.Operator.DOT_PRODUCT
+    trans = DotProduct2CodeTrans()
+    with pytest.raises(TransformationError) as info:
+        trans.validate(dot_product)
+    assert expected in str(info.value)
+
+
+def check_trans(code, expected, fortran_reader, fortran_writer, tmpdir,
+                index=0):
+    '''Utility function that takes fortran code and checks that the
+    expected code is produced when the DotProduct2CodeTrans
+    transformation is applied. The code is also checked to see if it
+    will compile.
+
+    :param str code: the tangent linear code stored in a string.
+    :param str expected: the expected adjoint code stored in a string.
+    :param fortran_reader: converts Fortran into PSyIR.
+    :type fortran_reader: :py:class:`psyclone.psyir.frontend.fortran`
+    :param fortran_writer: converts PSyIR into Fortran.
+    :type fortran_writer: :py:class:`psyclone.psyir.backend.fortran`
+    :param tmpdir: path to a test-specific temporary directory in \
+        which to test compilation.
+    :type tmpdir: :py:class:`py._path.local.LocalPath`
+    :param integer index: specifies which binaryoperation to check in \
+        code order. Defaults to 0 (which is the first).
+
+    '''
+    psyir = fortran_reader.psyir_from_source(code)
+    dot_product = psyir.walk(BinaryOperation)[index]
+    assert dot_product.operator == BinaryOperation.Operator.DOT_PRODUCT
+    trans = DotProduct2CodeTrans()
+    trans.apply(dot_product)
+    result = fortran_writer(psyir)
+    assert expected in result
+    assert Compile(tmpdir).string_compiles(result)
+
+
+# _get_array_bound function
+
+def test_bound_error(fortran_writer):
+    '''Test that the expected exception is raised if both arguments are
+    not references to arrays.
+
+    '''
+    with pytest.raises(TransformationError) as info:
+        _get_array_bound(Literal("1.0", REAL_TYPE), Literal("2.0", REAL_TYPE),
+                         fortran_writer)
+    assert ("DotProduct2CodeTrans._get_array_bound requires at least one of "
```

Comment:
In fact, we should throw an error if any argument is *not* a Reference. I think this relates to the comments I made in validate() earlier.

Reply:
This test has been removed as the offending exception has also been removed. This is because the exception was causing confusion and was not needed as the validate method checks for any errors.